cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -490,6 +577,60 @@ private void closeScanner() {
     }
   }
 
+  /**
+   * Filters elements matching an index {@link ScannerIterator}.
+   */
+  @VisibleForTesting
+  static final class FilterByIndexIterator extends AbstractCloseableIterator<StructuredRow> {
+    private final CloseableIterator<StructuredRow> scannerIterator;
+    private final Predicate<StructuredRow> predicate;
+
+    FilterByIndexIterator(CloseableIterator<StructuredRow> scannerIterator, Field<?> filterIndex,
+                          StructuredTableSchema schema) {
+      this.scannerIterator = scannerIterator;
+
+      switch (filterIndex.getFieldType()) {
+        case INTEGER:
+          predicate = row ->Objects.equals(row.getInteger(filterIndex.getName()), filterIndex.getValue());
+          break;
+        case LONG:
+          predicate = row ->Objects.equals(row.getLong(filterIndex.getName()), filterIndex.getValue());
+          break;
+        case FLOAT:
+          predicate = row ->Objects.equals(row.getFloat(filterIndex.getName()), filterIndex.getValue());
+          break;
+        case DOUBLE:
+          predicate = row ->Objects.equals(row.getDouble(filterIndex.getName()), filterIndex.getValue());
+          break;
+        case STRING:
+          predicate = row ->Objects.equals(row.getString(filterIndex.getName()), filterIndex.getValue());
+          break;
+        case BYTES:
+          predicate = row ->Arrays.equals(row.getBytes(filterIndex.getName()), (byte[]) filterIndex.getValue());
+          break;
+        default:
+          throw new InvalidFieldException(schema.getTableId(), filterIndex.getName());
+      }
+    }
+
+    @Override
+    protected StructuredRow computeNext() {
+      // Postfiltering on scanned rows by index
+      while (scannerIterator.hasNext()) {
```

Comment:
`while` --> `if`

Reply:
so this one has to be a while loop, as per the logic we return only when we get a match

Reply:
my bad, this one should stay as `while`